packages/talos_messenger_actions/src/kafka/producer.rs

```diff
@@ -0,0 +1,143 @@
+use std::collections::HashMap;
+
+use async_trait::async_trait;
+use log::{debug, info};
+use rdkafka::{
+    config::{FromClientConfig, FromClientConfigAndContext},
+    producer::{BaseRecord, DefaultProducerContext, ProducerContext, ThreadedProducer},
+    ClientContext, Message,
+};
+use talos_certifier::{
+    errors::SystemServiceError,
+    ports::{common::SharedPortTraits, errors::MessagePublishError, MessagePublisher},
+};
+use talos_certifier_adapters::kafka::utils::build_kafka_headers;
+use talos_rdkafka_utils::kafka_config::KafkaConfig;
+use tokio::sync::mpsc;
+
+use futures_executor::block_on;
+use talos_messenger_core::core::MessengerChannelFeedback;
+
+#[derive(Debug)]
+pub struct MessengerProducerDeliveryOpaque {
+    pub version: u64,
+    pub total_publish_count: u32,
+}
+
+pub struct MessengerKafkaProducerContext {
+    pub tx_feedback_channel: mpsc::Sender<MessengerChannelFeedback>,
+}
+
+impl ClientContext for MessengerKafkaProducerContext {}
+impl ProducerContext for MessengerKafkaProducerContext {
+    type DeliveryOpaque = Box<MessengerProducerDeliveryOpaque>;
+
+    fn delivery(&self, delivery_result: &rdkafka::producer::DeliveryResult<'_>, delivery_opaque: Self::DeliveryOpaque) {
+        let result = delivery_result.as_ref();
+
+        let version = delivery_opaque.version;
+
+        match result {
+            Ok(msg) => {
+                info!("Message {:?} {:?}", msg.key(), msg.offset());
+                // TODO: GK - what to do on error? Panic?
+                let _ = block_on(self.tx_feedback_channel.send(MessengerChannelFeedback::Success(
+                    version,
+                    "kafka".to_string(),
+                    delivery_opaque.total_publish_count,
+                )));
+            }
+            Err(err) => {
+                // TODO: GK - what to do on error? Panic?
```

Comment:
If we panic and do not commit offset then pods will keep restarting and will be stuck in that consume/panic/restart loop.

Reply:
Failure to commit, while undesirable, shouldn't violate safety.

Reply:
The `todo` is for the channel. What to do if the channel gives an error. As this is the way I get to know some Kafka action has been processed, and then mark it in suffix. 

Realistically the channel would give error when we try to send the message after the receiver was closed, so can I make the assumption that we can ignore the error check for `channel.send` ?